node/src/components/block_accumulator.rs

```diff
@@ -227,37 +245,65 @@ impl BlockAccumulator {
     ) {
-        let maybe_local_tip_era_id = self.local_tip.map(|local_tip| local_tip.era_id);
-        match (maybe_era_id, maybe_local_tip_era_id) {
-            // When the era of the item sent by this peer is known, we check it
-            // against our local tip era. If it is recent (within a number of
-            // eras of our local tip), we create an acceptor for it along with
-            // registering the peer.
-            (Some(era_id), Some(local_tip_era_id))
-                if era_id >= local_tip_era_id.saturating_sub(self.recent_era_interval) =>
+        // If the acceptor already exists, just register the peer, if applicable.
+        let entry = match self.block_acceptors.entry(block_hash) {
+            btree_map::Entry::Occupied(entry) => {
+                if let Some(sender) = maybe_sender {
+                    entry.into_mut().register_peer(sender);
+                }
+                return;
+            }
+            btree_map::Entry::Vacant(entry) => entry,
+        };
+
+        // The acceptor doesn't exist. Don't create it if the item's era is not provided, the local
+        // tip doesn't have an era or the item's era is older than the local tip era by more than
+        // `recent_era_interval`.
+        match (maybe_era_id, self.local_tip) {
+            (Some(era_id), Some(local_tip))
+                if era_id >= local_tip.era_id.saturating_sub(self.recent_era_interval) => {}
+            _ => return,
+        }
+
+        // Check that the sender isn't telling us about more blocks than expected.
+        if let Some(sender) = maybe_sender {
+            let block_timestamps = self.peer_block_timestamps.entry(sender).or_default();
+
+            // Prune the timestamps, so the count reflects only the most recently added acceptors.
+            let purge_interval = self.purge_interval;
+            while block_timestamps
+                .front()
+                .map_or(false, |(_, timestamp)| timestamp.elapsed() > purge_interval)
             {
-                let acceptor = self
-                    .block_acceptors
-                    .entry(block_hash)
-                    .or_insert_with(|| BlockAcceptor::new(block_hash, None));
-                acceptor.register_peer(sender);
+                block_timestamps.pop_front();
             }
-            // In all other cases (i.e. the item's era is not provided, the
-            // local tip doesn't have an era or the item's era is older than
-            // the local tip era by more than `recent_era_interval`), we only
-            // register the peer if there is an acceptor for the item already.
-            _ => {
-                if let Some(acceptor) = self.block_acceptors.get_mut(&block_hash) {
-                    acceptor.register_peer(sender)
-                }
+
+            // Assume a block time of at least 1 millisecond, so we don't divide by zero.
+            let block_time = self.block_time.max(TimeDiff::from(1));
```

Comment:
Quietly discarding bad values here is not optimal (i.e. if there's a bad block time configured).

We should probably ensure the `TimeDiff` is in range in the component constructor and return an error, or panic, or just use `TimeDiff::from(1)` there as a default, then rely on this property here.

Reply:
But in theory, a minimum delay of 0 is valid, or at least it will be in 2.0!
And then we don't really have an upper limit for the block per purge time. I'm not sure what to do in that case.